go/internal/services/usagelimiter/redis.go

```diff
@@ -0,0 +1,382 @@
+//   - Uses Redis INCRBY for atomic decrements (single operation)
+//   - No race conditions - Redis handles atomicity
+//   - Simple, fast code with minimal complexity
+//   - Excellent performance under high contention
+//   - Falls back to direct DB on Redis failures
+//
+// Parameters:
+//   - config: Configuration options including counter implementation
+//
+// Returns:
+//   - Service: Counter-based implementation of the Service interface
+//   - error: Any initialization errors
+func NewCounter(config CounterConfig) (Service, error) {
+	// Validate required fields
+	if config.Counter == nil {
+		return nil, fmt.Errorf("counter cannot be nil")
+	}
+	if config.DB == nil {
+		return nil, fmt.Errorf("database cannot be nil")
+	}
+
+	ttl := config.TTL
+	if ttl == 0 {
+		ttl = defaultTTL
+	}
+
+	// Create the direct DB fallback service
+	dbFallback, err := New(Config{
+		DB:     config.DB,
+		Logger: config.Logger,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to create DB fallback: %w", err)
+	}
+
+	s := &counterService{
+		db:         config.DB,
+		logger:     config.Logger,
+		counter:    config.Counter,
+		dbFallback: dbFallback,
+		ttl:        ttl,
+		replayBuffer: buffer.New[CreditChange](buffer.Config{
+			Name:     "usagelimiter_replays",
+			Capacity: 10_000,
+			Drop:     false, // Do NOT drop credits
+		}),
+		dbCircuitBreaker: circuitbreaker.New[any](
+			"usagelimiter_db_writes",
+			circuitbreaker.WithLogger(config.Logger),
+		),
+	}
+
+	// Start replay workers
+	replayWorkers := config.ReplayWorkers
+	if replayWorkers <= 0 {
+		replayWorkers = defaultReplyWorkers
+	}
+
+	for range replayWorkers {
+		go s.replayRequests()
+	}
+
+	return s, nil
+}
+
+// Limit processes a usage request and enforces credit limits using distributed Redis counters.
+//
+// This function implements a two-phase credit limiting system:
+// 1. When key exists in redis: Atomic decrement on existing Redis counters
+// 2. When key does not exist in redis: Database load + Redis initialization for new keys
+// ERROR HANDLING:
+// - Redis failures trigger automatic fallback to direct DB operations
+// - Lua script prevents negative counters, insufficient credit attempts are denied atomically
+// - Circuit breaker protects against database overload during fallbacks
+//
+// METRICS:
+// - usagelimiter_decisions_total: Tracks allowed/denied decisions by backend
+// - usagelimiter_credits_processed_total: Tracks total credits consumed
+// - usagelimiter_fallback_operations_total: Tracks fallback frequency
+func (s *counterService) Limit(ctx context.Context, req UsageRequest) (UsageResponse, error) {
+	ctx, span := tracing.Start(ctx, "usagelimiter.counter.Limit")
+	defer span.End()
+
+	redisKey := fmt.Sprintf("credits:%s", req.KeyId)
+
+	// Try decrement only if key usage is already stored
+	remaining, exists, err := s.counter.DecrementIfExists(ctx, redisKey, int64(req.Cost))
+	if err != nil {
+		metrics.UsagelimiterFallbackOperations.Inc()
```

Comment:
nit: maybe this should be in the dbFallback.Limit method

it's odd that in this case we have a metric here, but in the other two cases below we do not.